src/App.tsx

```diff
@@ -1,26 +1,161 @@
+import { TodoList } from './components/TodoList';
+import { TodoFilter } from './components/TodoFilter';
+import { ErrorNotification } from './components/ErrorNotification';
+import { Header } from './components/Header';
+import { Status } from './types/Status';
+import { Todo } from './types/Todo';
+import { USER_ID, deleteTodo, getTodos, updateTodo } from './api/todos';
+import { isAllTodosCompleted, getAllActiveTodos } from './utils/finder';
+import { ErrorMessages } from './types/ErrorMessages';
+import { filterTodosByStatus } from './utils/filterTodos';
 
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [status, setStatus] = useState<Status>(Status.All);
+  const [errorMessage, setErrorMessage] = useState<string | null>(null);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [processingTodos, setProcessingTodos] = useState<number[]>([]);
+
+  const fieldTitle = useRef<HTMLInputElement>(null);
+
+  const isAllCompleted = useMemo(() => isAllTodosCompleted(todos), [todos]);
+  const totalTodosActive = useMemo(() => getAllActiveTodos(todos), [todos]);
+
+  useEffect(() => {
+    getTodos()
+      .then(setTodos)
+      .catch(() => setErrorMessage(ErrorMessages.LOAD_TODOS));
+  }, []);
+
+  const filteredTodos = useMemo(
+    () => filterTodosByStatus(todos, status),
+    [status, todos],
+  );
+
   if (!USER_ID) {
     return <UserWarning />;
   }
 
+  const onAddTodo = (todo: Todo) => {
+    setTodos(prevTodos => [...prevTodos, todo]);
+  };
+
+  const onDeleteTodo = (id: number) => {
+    setProcessingTodos([id]);
+
+    deleteTodo(id)
+      .then(() => {
+        setTodos(prevTodos => prevTodos.filter(todo => todo.id !== id));
+        setProcessingTodos([]);
+      })
+      .catch(() => setErrorMessage(ErrorMessages.DELETE_TODO))
+      .finally(() => {
+        fieldTitle.current?.focus();
+      });
+  };
+
+  const onDeleteCompleted = () => {
+    Promise.allSettled(
+      todos
+        .filter(todo => todo.completed)
+        .map(todo => {
+          setProcessingTodos(prev => [...prev, todo.id]);
+
+          return deleteTodo(todo.id)
+            .then(() =>
+              setTodos(prevTodos =>
+                prevTodos.filter(prevTodo => prevTodo.id !== todo.id),
+              ),
+            )
+            .catch(() => setErrorMessage(ErrorMessages.DELETE_TODO));
+        }),
+    ).then(() => {
+      fieldTitle.current?.focus();
+      setProcessingTodos([]);
+    });
+  };
+
+  const onTodoUpdate = (updatedTodo: Todo) => {
+    setProcessingTodos(prevId => [...prevId, updatedTodo.id]);
+
+    return updateTodo(updatedTodo)
+      .then(todo => {
+        setTodos(prevTodos =>
+          prevTodos.map(prevTodo =>
+            prevTodo.id === todo.id ? updatedTodo : prevTodo,
+          ),
+        );
+      })
+      .catch(() => {
+        setErrorMessage(ErrorMessages.UPDATE_TODO);
+        throw new Error();
```

Comment:
do not throw an error here - you handled it already, so no need to throw it again. It will lead to a page crash (because you don't try to catch the error thrown here somewhere else)